test/unit/Cardano/Wallet/BlockSyncerSpec.hs

```diff
@@ -0,0 +1,177 @@
+        loop blockNumToGo res = do
+            let bytelistGenerator = pack <$> vector 10 :: Gen ByteString
+            if blockNumToGo <= 0 then
+                return $ reverse res
+            else
+               case res of
+                   [] -> do
+                       lastBlockHeader <- Hash <$> generate bytelistGenerator
+                       theBlockHeader <- Hash <$> generate bytelistGenerator
+                       let theEpoch = 0
+                       let theSlot = 1
+                       let theBlock = Block (BlockHeader theEpoch theSlot lastBlockHeader) Set.empty
+                       loop (blockNumToGo - 1) [(theBlockHeader, theBlock)]
+                   (lastBlockHeader, Block (BlockHeader lastEpoch lastSlot _) _ ):_ -> do
+                       let theBlock = Block (BlockHeader lastEpoch (lastSlot + 1) lastBlockHeader) Set.empty
+                       theBlockHeader <- Hash <$> generate bytelistGenerator
+                       loop (blockNumToGo - 1) $ (theBlockHeader, theBlock):res
+
+
+newtype BlocksConsumed = BlocksConsumed [(Hash "BlockHeader")] deriving (Show, Eq)
+
+newtype BlocksToInject = BlocksToInject ([Int],[((Hash "BlockHeader"),Block)]) deriving (Show, Eq)
+
+
+data DeliveryMode = ExactlyOnce | AtLeastOnce
+
+pushNextBlocks
+    :: MVar BlocksToInject
+    -> DeliveryMode
+    -> IO [Block]
+pushNextBlocks ref mode = do
+    (BlocksToInject (blocksToTake, blocksRemaining)) <- takeMVar ref
+    if L.null blocksRemaining then return [] else
+        (do case blocksToTake of
+                num : rest -> do
+                    let (bOut, bStay) = L.splitAt num blocksRemaining
+                    putMVar ref $ BlocksToInject (rest, bStay)
+                    case mode of
+                        ExactlyOnce ->
+                            return $ map snd bOut
+                        AtLeastOnce -> do
+                            additionalBlocks <- generate $ choose (1,3) :: IO Int
+                            return $ map snd (bOut ++ take additionalBlocks bStay)
+                [] -> return [])
+
+
+writeToIORefAction
+    :: MVar BlocksConsumed
+    -> [((Hash "BlockHeader"),Block)]
+    -> IO (Block -> IO ())
+writeToIORefAction ref blocks = return $
+    \block -> do
+        (BlocksConsumed headerHashesConsumed) <- takeMVar ref
+        case (map fst . filter (\(_,b) -> b == block) ) blocks of
+            [blockHeader] -> do
+                putMVar ref $ BlocksConsumed (blockHeader : headerHashesConsumed)
+            _ ->
+                return ()
+
+
+generateBlockChunkSizes :: IO [Int]
+generateBlockChunkSizes = do
+    numberOfTicks <- generate $ choose (1,15)
+    generate $ generateBlockChunks numberOfTicks
+        where
+            generateBlockChunks
+                :: Int
+                -> Gen [Int]
+            generateBlockChunks numberOfTicks = do
+                let chunkSizeGen = choose (0,15)
+                vectorOf numberOfTicks chunkSizeGen
```

Comment:
It would have been nice to have a chat about the approach on testing for this. I think you've got the idea, but the implementation of the tests has gone slightly too complex IMO. I'll make a suggestion a bit later but it would be nice to go through this together.